### PR TITLE
add `cast-nan-to-int` lint

### LINKED ISSUE #371: Casting a known NaN value to an integral type
This is actually undefined behavior, and doesn't even get a warning from rustc.

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #371.

r? @Alexendoo

---

changelog: add [`cast-nan-to-int`] lint

### CODE REVIEW

File: tests/ui/cast_nan_to_int.stderr [hunk 1]
```diff
@@ -0,0 +1,40 @@
+error: casting a known NaN to usize
+  --> $DIR/cast_nan_to_int.rs:5:13
+   |
+LL |     let _ = (0.0_f32 / -0.0) as usize;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: `-D clippy::cast-nan-to-int` implied by `-D warnings`
```
Comment (Alexendoo) [inline tests/ui/cast_nan_to_int.stderr:7]: Would be worth adding a note explaining that the result will be 0

Comment (Alexendoo) [reply tests/ui/cast_nan_to_int.stderr:7]: Maybe we should just lint all operations that result in `NaN`? Whether cast or not

Comment (llogiq) [reply tests/ui/cast_nan_to_int.stderr]: That would mean we'd need to check far more expressions. This way we only look into the constant evaluation when we have a cast from float to int, which is sufficiently rare (also in earlier versions of Rust, this cast was actually UB, but that's quite a while ago now). Also the risk of false positives seems higher.

I'm fine with retiring this lint once we have one that checks for all NaNs and works reasonably well.

**Comment by llogiq** [comment]
I've added the note you suggested. Anything else needed?

**Review by Alexendoo — APPROVED** [review]
Nothing else, looks good!

you can r=me it with the conflict resolved

**Comment by llogiq** [comment]
Thank you for the review!

@bors r=@Alexendoo

**Pull request merged**